src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -50,18 +51,174 @@
+      case 0:
+        grpc_pollset_set_del_fd(ac->interested_parties_, fd);
+        *ep = grpc_tcp_client_create_from_fd(fd, ac->channel_args_,
+                                             ac->addr_str_);
+        fd = nullptr;
+        break;
+      case ENOBUFS:
+        /* We will get one of these errors if we have run out of
+           memory in the kernel for the data structures allocated
+           when you connect a socket.  If this happens it is very
+           likely that if we wait a little bit then try again the
+           connection will work (since other programs or this
+           program will close their network connections and free up
+           memory).  This does _not_ indicate that there is anything
+           wrong with the server we are connecting to, this is a
+           local problem.
+
+           If you are looking at this code, then chances are that
+           your program or another program on the same computer
+           opened too many network connections.  The "easy" fix:
+           don't do that! */
+        gpr_log(GPR_ERROR, "kernel out of buffers");
+        ac->mu_.Unlock();
+        grpc_fd_notify_on_write(fd, &ac->write_closure_);
+        return;
+      case ECONNREFUSED:
+        /* This error shouldn't happen for anything other than connect(). */
+        error = GRPC_OS_ERROR(so_error, "connect");
+        break;
+      default:
+        /* We don't really know which syscall triggered the problem here,
+           so punt by reporting getsockopt(). */
+        error = GRPC_OS_ERROR(so_error, "getsockopt(SO_ERROR)");
+        break;
+    }
+
+  finish:
+    if (fd != nullptr) {
+      grpc_pollset_set_del_fd(ac->interested_parties_, fd);
+      grpc_fd_orphan(fd, nullptr, nullptr, "tcp_client_orphan");
+      fd = nullptr;
+    }
+    done = (--ac->refs_ == 0);
+    ac->mu_.Unlock();
+    if (error != GRPC_ERROR_NONE) {
+      std::string str;
+      bool ret = grpc_error_get_str(error, GRPC_ERROR_STR_DESCRIPTION, &str);
+      GPR_ASSERT(ret);
+      std::string description =
+          absl::StrCat("Failed to connect to remote host: ", str);
+      error =
+          grpc_error_set_str(error, GRPC_ERROR_STR_DESCRIPTION, description);
+      error =
+          grpc_error_set_str(error, GRPC_ERROR_STR_TARGET_ADDRESS, addr_str);
+    }
+    if (done) {
+      delete ac;
+    }
+    // Push async connect closure to the executor since this may actually be
+    // called during the shutdown process, in which case a deadlock could form
+    // between the core shutdown mu and the connector mu (b/188239051)
+    grpc_core::Executor::Run(closure, error);
+  }
+
+  grpc_core::Mutex mu_;
+  grpc_fd* fd_ ABSL_GUARDED_BY(&mu_);
+  grpc_timer alarm_ ABSL_GUARDED_BY(&mu_);
+  grpc_closure on_alarm_ ABSL_GUARDED_BY(&mu_);
+  int refs_ ABSL_GUARDED_BY(&mu_);
+  grpc_closure write_closure_ ABSL_GUARDED_BY(&mu_);
+  grpc_pollset_set* interested_parties_ ABSL_GUARDED_BY(&mu_);
+  std::string const addr_str_;
```

Comment:
stylistically, the rest of the codebase puts `const` before the type. See "Where to put the const" in https://google.github.io/styleguide/cppguide.html#Use_of_const. Same for the types below.